### PR TITLE
Add analytics event for extract column in chill mode

### LINKED ISSUE #41353: [FE] Add `extract_via_column_header` analytics event
`extract_via_column_header`

https://www.notion.so/metabase/0df8245abb8f46edbede12065d65e774?v=86c486a9c8904fb0ab086ad389f63f81

https://metaboat.slack.com/archives/C0645JP1W81/p1712845418159769

----

update event status in the doc once it's ready

----

This task should really go into #38964, but that epic has already been closed

### LINKED ISSUE #41773: Add an analytics event for extract column in the chill mode header
Add an [analytics event](https://www.notion.so/metabase/Part-of-column-is-extracted-via-column-header-button-6575d283efe542408781eb42cbaedc3d) for the extract column action in the chill mode header.

### PR DESCRIPTION
**Pull request by romeovs**
Closes https://github.com/metabase/metabase/issues/41773

### Description

Adds an analytics event that fires when the custom column is created from the table mode via Extract button in the column header

### CODE REVIEW

**Comment by romeovs** [comment]
Superseded by https://github.com/metabase/metabase/pull/41767

**Review by ranquild — CHANGES_REQUESTED** [review]
`structEvents` are deprecated google analytics events that should not be used. Please use snowplow events.

File: frontend/src/metabase/querying/analytics.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  trackSchemaEvent("question", "1-0-4", {
+    event: "column_extract_via_column_header",
+    custom_expressions_used: expressionsUsedBy(tag),
+    database_id: Lib.databaseID(query),
+    question_id: question?.id() ?? 0,
+  });
+};
+
+function expressionsUsedBy(tag: string) {
+  switch (tag) {
```
Comment (romeovs) [inline frontend/src/metabase/querying/analytics.ts:19]: This is a bit fragile: when we add more extractions, we need to add the expressions used here too.

But AFAIC there is no easy way to get the expressions introduced by an extraction from the extraction itself.

Comment (ranquild) [reply frontend/src/metabase/querying/analytics.ts:19]: We could make `ColumnExtractionInfo.tag` to be a enum and not just a string. We could also define a mapping type used here in a way to make it fail if there is a new tag that is not supported here. Like:
```
Record<ColumnExtractionTag, string>
```

Another way could to be to make analytic events match the expression tag either by updating the tag or the analytics spec.

**Review by ranquild — COMMENTED** [review]
Inline comments.

File: frontend/src/metabase/querying/analytics.ts (hunk 1 above)
Comment (romeovs) [reply frontend/src/metabase/querying/analytics.ts]: Ok I've done exactly that, please have another look.

**Comment by ranquild** [comment]
@romeovs same as https://github.com/metabase/metabase/pull/42095#issuecomment-2088397865

**Comment by romeovs** [comment]
@ranquild please have another look, Braden added a helper so we could actually get the extractions from the drill.

That is used to get the functions from the expression generated by the extraction correctly.

File: frontend/src/metabase-lib/extractions.ts [hunk 2]
```diff
@@ -9,10 +36,39 @@ export function columnExtractions(
...
+  return walk(parts);
+}
+
+function walk(parts: ExpressionParts): string[] {
+  const res: string[] = [parts.operator];
+  parts.args.forEach(arg => {
+    if (!arg || !(typeof arg === "object")) {
+      return;
+    }
+    if ("operator" in arg) {
```
Comment (ranquild) [inline frontend/src/metabase-lib/extractions.ts:69]: `if (`typeof arg === "object") walk(...)` without the first if

**Review by ranquild — APPROVED** [review]

File: frontend/src/metabase-lib/extractions.ts (hunk 2 above)
Comment (ranquild) [reply frontend/src/metabase-lib/extractions.ts:69]: https://github.com/metabase/metabase/blob/4b71e9aa645117a944af409882bda33f64225bf9/frontend/src/metabase-lib/filter.ts#L570

Comment (ranquild) [reply frontend/src/metabase-lib/extractions.ts:69]: because `typeof null === "object"`

Comment (ranquild) [reply frontend/src/metabase-lib/extractions.ts:69]: feel free to move this utility to a different place

**Pull request merged**